src/Features/Core/Portable/CustomMessageHandler/CustomMessageHandlerService.cs

```diff
@@ -0,0 +1,350 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#if !NETSTANDARD2_0
+
+using System;
+using System.Collections.Generic;
+using System.Composition;
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Runtime.Loader;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Internal.Log;
+
+namespace Microsoft.CodeAnalysis.CustomMessageHandler;
+
+[ExportWorkspaceService(typeof(ICustomMessageHandlerService)), Shared]
+internal sealed class CustomMessageHandlerService : ICustomMessageHandlerService, IDisposable
+{
+    /// <summary>
+    /// Extensions assembly load contexts and loaded handlers, indexed by handler file path. The handlers are indexed by type name.
+    /// </summary>
+    private readonly Dictionary<string, CustomMessageHandlerExtension> _extensions = new();
+
+    /// <summary>
+    /// Handlers of document-related messages, indexed by handler message name.
+    /// </summary>
+    private readonly Dictionary<string, ICustomMessageDocumentHandlerWrapper> _documentHandlers = new();
+
+    /// <summary>
+    /// Handlers of non-document-related messages, indexed by handler message name.
+    /// </summary>
+    private readonly Dictionary<string, ICustomMessageHandlerWrapper> _handlers = new();
+
+    private readonly ICustomMessageHandlerFactory _customMessageHandlerFactory;
+
+    private readonly object _lockObject = new();
+
+    private readonly AssemblyLoadContext _defaultLoadContext
+        = AssemblyLoadContext.GetLoadContext(typeof(CustomMessageHandlerService).Assembly)
+        ?? throw new InvalidOperationException($"Cannot get assembly load context for {nameof(CustomMessageHandlerService)}.");
+
+    [ImportingConstructor]
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public CustomMessageHandlerService(ICustomMessageHandlerFactory customMessageHandlerFactory)
+    {
+        _customMessageHandlerFactory = customMessageHandlerFactory;
+    }
+
+    public ValueTask<RegisterHandlersResponse> LoadCustomMessageHandlersAsync(
+        string assemblyFolderPath,
+        string assemblyFileName,
+        CancellationToken cancellationToken)
+    {
+#if DEBUG
+        System.Diagnostics.Debugger.Launch();
+#endif
```

Comment:
this seems like soemthing we'll need to remove?

Reply:
Yes, I will remove all of these as soon as I get a go-ahead with this implementation and I know I won't have to do so much debugging of code changes :-)